fabric-api-base/src/main/java/net/fabricmc/fabric/api/event/Event.java

```diff
@@ -43,6 +43,17 @@ public final T invoker() {
 		return invoker;
 	}
 
+	/**
+	 * Some events can be optimized away if nobody is listening. If an Event
+	 * implementation can quickly check for this condition, it may override this
+	 * method to return false in that case.
+	 *
+	 * @return True if the event might have any listeners.
+	 */
+	public boolean hasListeners() {
```

Comment:
I don't know how this works out with the concurrency-compatible design of fab events (that's why invokers are a field because player wants you to update it real time than through some delay). But such logic usually have problems in a concurrent setting, where you are supposed to just try do an action and see its results than ask if you can do and then do an action (during the ask and do something else can meddle)

Reply:
If I recall player does do COW for adding new listeners in the events.

Reply:
Yes, I believe he does that though generating new invokers through the function supplied to `createArrayBacked`, which is an array to created invoker function. In an invocation the old invoker can be removed from the event instance but its invocation will continue, which is effectively copy-on-write.

Reply:
I don't see this as an issue right now because `ArrayBackedEvent` already has a race condition anyway, the `hasListeners` only makes the race window slightly bigger, might need to be documented but not sure about refactoring. Notice the small window between [updating the handlers](https://github.com/FabricMC/fabric/blob/ed0c348cc4cda9cff8abc5953b370fa872cba943/fabric-api-base/src/main/java/net/fabricmc/fabric/impl/base/event/ArrayBackedEvent.java#L64-L71) and [changing the invoker](https://github.com/FabricMC/fabric/blob/ed0c348cc4cda9cff8abc5953b370fa872cba943/fabric-api-base/src/main/java/net/fabricmc/fabric/impl/base/event/ArrayBackedEvent.java#L49). If an `Event` implementation _really_ wants to be as concurrent as possible it doesn't even have to implement this function anyway. This approach can only ever extend the window if an invoker is added in between the call to `Event#hasListeners()` and the call to `Event#invoker()`. As with the `dummyInvoker` optimization in `ArrayBackedEvent` or the COW, there really is no way to do this optimization without creating a race condition. Do we want to optimize away ~100 bytecode instructions for every event invocation (If no-one is listening), or do we want event listeners to start being able to receive events ~100 bytecodes earlier? It doesn't seem like a big deal to me to not invoke listeners for the sake of optimization for events that were emitted before the listener was even added.

Another option I considered was to instead expose a `T Event#noop()` that the `invoker` can be compared against (Would return `this.dummyHandler` in `ArrayBackedEvent`), this doesn't change the race condition in any way, but it might make it more obvious that the `invoker()` is slightly out of date by the time it gets invoked. I preferred the boolean approach purely for performance reasons because it doesn't force the event implementation to cache a `noop` handler if it wants to have this optimization.